app/src/testDebug/java/com/futurice/freesound/feature/common/ui/adapter/RecyclerViewAdapterTest.java

```diff
@@ -0,0 +1,212 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.futurice.freesound.feature.common.ui.adapter;
+
+import com.futurice.freesound.feature.common.DisplayableItem;
+import com.futurice.freesound.feature.common.scheduling.SchedulerProvider;
+import com.futurice.freesound.test.rx.TrampolineSchedulerProvider;
+import com.futurice.freesound.viewmodel.viewholder.BaseBindingViewHolder;
+
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.MockitoAnnotations;
+
+import android.support.v7.widget.RecyclerView;
+import android.view.ViewGroup;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.atLeast;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
+
+/**
+ * Created by Futurice
+ */
+public class RecyclerViewAdapterTest {
+
+    @Mock
+    private ItemComparator comparator;
+
+    @Mock
+    private ViewHolderFactory factory1;
+
+    @Mock
+    private ViewHolderFactory factory2;
+
+    @Mock
+    private ViewHolderBinder binder1;
+
+    @Mock
+    private ViewHolderBinder binder2;
+
+    @Rule
+    public ExpectedException thrown = ExpectedException.none();
+
+    private RecyclerViewAdapter adapter;
+
+    @Before
+    public void setUp() {
+        MockitoAnnotations.initMocks(this);
+
+        SchedulerProvider schedulerProvider = new TrampolineSchedulerProvider();
+        adapter = new RecyclerViewAdapter(comparator, factoryMap(), binderMap(), schedulerProvider);
+    }
+
+    @Test
+    public void expectedNPEWhenViewGroupIsNullInOnCreateViewHolder() {
+        thrown.expect(NullPointerException.class);
+
+        adapter.onCreateViewHolder(null, 1);
+    }
+
+    @Test
+    public void expectedNPEWhenTypeDoesNotExistInOnCreateViewHolder() {
+        thrown.expect(NullPointerException.class);
+
+        adapter.onCreateViewHolder(null, 5);
+    }
+
+    @Test
+    public void factoryForPassedTypeCreatesViewHolder() {
+        ViewGroup viewGroup = mock(ViewGroup.class);
+
+        adapter.onCreateViewHolder(viewGroup, 1);
+
+        Mockito.verify(factory1).createViewHolder(viewGroup);
+    }
+
+    @Test
+    public void expectedNPEWhenNullIsPassedToUpdate() {
+        thrown.expect(NullPointerException.class);
+
+        adapter.update(null);
+    }
+
+    @Test
+    public void diffUtilsIsNotUsedWhenUpdatingAdapterForTheFirstTime() {
+        List<DisplayableItem> itemList = itemList();
+
+        // FIXME: AndroidPreconditions.assertUiThread(), how to mock this ?
+        adapter.update(itemList);
```

Comment:
The `SchedulerProvider` has a method `isUiThread` which mockable. The assertion in the Adapter could be changed to use that, rather than `AndroidPreconditions`.